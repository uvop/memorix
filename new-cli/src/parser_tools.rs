use nom::combinator::cut;
use nom::error::context;
use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{alpha1, alphanumeric1, char, multispace0, multispace1},
    combinator::{map, recognize},
    error::ParseError,
    multi::{many0, separated_list0},
    sequence::{pair, preceded, separated_pair, terminated, tuple},
    IResult,
};

pub fn indent(level: usize) -> String {
    "  ".repeat(level)
}

#[macro_export]
macro_rules! permutation_many {
    () => {
        many_m_n(0, 0, nom::combinator::fail)
    };
    ( $total:tt, $( ($parser:expr, $index:tt, $required:tt) ),+ $(,)? ) => {
        map(
            context(
                "Missing some required parsers here",
                verify(
                    map(
                        many_m_n(0, $total,
                            alt((
                                $( map($parser, move |x| {
                                    let mut tuple = permutation_many!(@create_none_tuple $total);
                                    tuple.$index = Some(x);
                                    tuple
                                }), )+
                            ))
                        ),
                        |vec| {
                            let tuple = vec.into_iter().fold(
                                permutation_many!(@create_none_tuple $total),
                                |mut acc, item| {
                                    $( permutation_many!(@collect $index, acc, item); )+
                                    acc
                                }
                            );
                            (
                                $( tuple.$index ),+
                            )
                        }
                    ),
                    #[allow(unused_variables)]
                    |tuple| {
                        $( permutation_many!(@verify_field tuple.$index, $required) && )+ true
                    }
                ),
            ),
            |tuple| {
                (
                    $( permutation_many!(@handle_required tuple.$index, $required), )+
                )
            }
        )
    };
    (@create_none_tuple 1) => { (None,) };
    (@create_none_tuple 2) => { (None, None) };
    (@create_none_tuple 3) => { (None, None, None) };
    (@create_none_tuple 4) => { (None, None, None, None) };
    (@create_none_tuple 5) => { (None, None, None, None, None) };
    (@create_none_tuple 6) => { (None, None, None, None, None, None) };
    (@collect $index:tt, $acc:expr, $item:expr) => {
        if $acc.$index.is_none() { $acc.$index = $item.$index; }
    };
    (@verify_field $field:expr, true) => {
        $field.is_some()
    };
    (@verify_field $field:expr, false) => {
        true
    };
    (@handle_required $field:expr, true) => {
        $field.expect("Shouldn't get here")
    };
    (@handle_required $field:expr, false) => {
        $field
    };
}

#[macro_export]
macro_rules! create_enum_with_const_slice {
    (
        $(#[$meta:meta])*
        $vis:vis enum $name:ident {
            $($(#[$variant_meta:meta])* $variant:ident),+ $(,)?
        }
        $const_vis:ident $const_name:ident
    ) => {
        $(#[$meta])*
        $vis enum $name {
            $($(#[$variant_meta])* $variant),+
        }

        $const_vis const $const_name: &'static [$name] = &[
            $($name::$variant),+
        ];
    }
}

pub trait FromSdl {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized;
}
pub trait ToSdl {
    fn to_sdl(&self, level: usize) -> String;
}

impl FromSdl for String {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            recognize(pair(
                alt((alpha1, tag("_"))),
                many0(alt((alphanumeric1, tag("_")))),
            )),
            |x: &str| x.to_string(),
        )(input)
    }
}

impl ToSdl for String {
    fn to_sdl(&self, _: usize) -> String {
        self.to_string()
    }
}

impl<T: FromSdl> FromSdl for Vec<T> {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        context(
            "array",
            preceded(
                tuple((char('['), multispace0)),
                cut(terminated(
                    separated_list0(multispace1, T::from_sdl),
                    preceded(multispace0, char(']')),
                )),
            ),
        )(input)
    }
}

impl<T: ToSdl> ToSdl for Vec<T> {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level + 1);
        let mut result = String::from("[\n");

        for item in self {
            result.push_str(&format!("{}{}\n", level_indent, item.to_sdl(level + 1)));
        }
        result.push_str(&format!("{}]", indent(level)));

        result
    }
}

impl<T: FromSdl> FromSdl for Vec<(String, T)> {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        context(
            "hash",
            preceded(
                char('{'),
                cut(terminated(
                    map(
                        separated_list0(
                            multispace1,
                            separated_pair(
                                preceded(multispace0, String::from_sdl),
                                cut(tuple((multispace0, char(':')))),
                                preceded(multispace0, T::from_sdl),
                            ),
                        ),
                        |tuple_vec| tuple_vec.into_iter().map(|(k, v)| (k, v)).collect(),
                    ),
                    preceded(multispace0, char('}')),
                )),
            ),
        )(input)
    }
}

impl<T: ToSdl> ToSdl for Vec<(String, T)> {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level + 1);
        let mut result = String::from("{\n");

        for (key, value) in self {
            result.push_str(&format!(
                "{}{}: {}\n",
                level_indent,
                key,
                value.to_sdl(level + 1)
            ));
        }
        result.push_str(&format!("{}}}", indent(level)));

        result
    }
}

#[macro_export]
macro_rules! impl_from_and_to_sdl_for_enum {
    ($enum_name:ident, $($variant:ident => $str_val:expr),+ $(,)?) => {
        impl FromSdl for $enum_name {
            fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(input: &'a str) -> IResult<&'a str, Self, E>
            where
                Self: Sized,
            {
                alt((
                    $(value($enum_name::$variant, tag($str_val))),+
                ))(input)
            }
        }
        impl ToSdl for $enum_name {
            fn to_sdl(&self, _: usize) -> String {
                match self {
                    $($enum_name::$variant => $str_val.to_string()),+
                }
            }
        }
    };
}

#[macro_export]
macro_rules! impl_from_and_to_sdl_for_struct {
    (($struct_name:ident, $total: tt), $(($field:ident: $field_type:ty, $index:tt, $required:tt)),+ $(,)?) => {
        impl FromSdl for $struct_name {
            fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(input: &'a str) -> IResult<&'a str, Self, E>
            where
                Self: Sized,
            {
                preceded(
                    char('{'),
                    cut(terminated(
                        permutation_many!(
                            $total,

                            $(
                                (
                                    preceded(
                                        tuple((multispace0, tag(stringify!($field)), multispace0, char(':'), multispace0)),
                                        cut(<$field_type>::from_sdl)
                                    ),
                                    $index,
                                    $required
                                ),
                            )+
                        ),
                        preceded(multispace0, char('}'))
                    ))
                )(input)
                .map(|(remaining, fields)| {
                    let ($($field),+) = fields;
                    (remaining, $struct_name { $($field),+ })
                })
            }
        }

        impl ToSdl for $struct_name {
            fn to_sdl(&self, level: usize) -> String {
                let level_indent = indent(level + 1);
                let mut result = String::from("{\n");

                $(
                    impl_from_and_to_sdl_for_struct!(@to_sdl_field self.$field, $field, $required, level_indent, result, level);
                )+

                result.push_str(&format!("{}}}", indent(level)));
                result
            }
        }
    };

    (@from_sdl_field $field:ident, $parser:expr, false) => {
        opt($parser)
    };

    (@from_sdl_field $field:ident, $parser:expr, true) => {
        context(concat!("Missing required key \"", stringify!($field), "\""), $parser)
    };

    (@to_sdl_field $field:expr, $field_name:ident, false, $indent:expr, $result:expr, $level:expr) => {
        if let Some(value) = &$field {
            $result.push_str(&format!(
                concat!("{}", stringify!($field_name),": {}\n"),
                $indent,
                value.to_sdl($level + 1)
            ));
        }
    };

    (@to_sdl_field $field:expr, $field_name:ident, true, $indent:expr, $result:expr, $level:expr) => {
        $result.push_str(&format!(
            concat!("{}", stringify!($field_name),": {}\n"),
            $indent,
            $field.to_sdl($level + 1)
        ));
    };
}
