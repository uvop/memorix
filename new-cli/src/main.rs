mod export_schemas;
mod flat_schema;
mod imports;
mod parser;
mod parser_tools;

use std::error::Error;
use std::fmt;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

use nom::error::convert_error;
use nom::Finish;
use parser::{schema_to_sdl, Schema};
use parser_tools::FromSdl;

pub struct PrettyError<T>(pub T);

impl<T> From<T> for PrettyError<T> {
    fn from(v: T) -> Self {
        Self(v)
    }
}

impl<T: fmt::Display> fmt::Debug for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl From<Box<dyn Error>> for PrettyError<String> {
    fn from(error: Box<dyn Error>) -> Self {
        PrettyError(error.to_string())
    }
}

impl<T: fmt::Debug + fmt::Display> Error for PrettyError<T> {}

impl<T: fmt::Display> fmt::Display for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String>;
    fn write_string(&self, path: &str, content: &str) -> io::Result<()>;
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>>;
}

pub struct RealFileSystem;

impl FileSystem for RealFileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String> {
        let mut file = File::open(path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        Ok(contents)
    }
    fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>> {
        let path = PathBuf::from(relative_file);
        let new_path = path
            .parent()
            .ok_or("Couldn't get parent of schema".to_string())?
            .join(target_file);
        let abs = new_path.canonicalize()?;
        Ok(abs
            .to_str()
            .ok_or("Couldn't convert path to string")?
            .to_string())
    }
}

pub fn format<F: FileSystem>(
    fs: &F,
    path: &str,
) -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let input = fs.read_to_string(path)?;
    let input = input.as_str();
    let (_, schema) = Schema::from_sdl(input).finish().map_err(|e| {
        let stack = format!("parser feedback:\n{}", convert_error(input, e));
        stack
    })?;
    let formatted_input = schema_to_sdl(&schema);
    fs.write_string(path, &formatted_input)?;
    Ok(())
}

pub fn codegen<F: FileSystem>(fs: &F, path: &str) -> Result<String, PrettyError<String>> {
    let parsed_schema =
        crate::imports::ImportedSchema::new(fs, path).map_err(|x| format!("{}", x))?;
    let export_schemas = crate::export_schemas::ExportSchema::new_vec(parsed_schema);
    let flat_export_schemas = export_schemas
        .into_iter()
        .map(crate::flat_schema::FlatExportSchema::new)
        .collect::<Vec<_>>();
    Ok(serde_json::to_string_pretty(&flat_export_schemas).unwrap())
}

enum Command {
    Format,
    Codegen,
}

fn main() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        return Err(format!("Usage: [codegen/format] <path_to_sdl_file>").into());
    }
    let command = match args[1].as_str() {
        "format" => Ok(Command::Format),
        "codegen" => Ok(Command::Codegen),
        x => Err(format!("Unknown command \"{}\"", x)),
    }?;

    let file_path = &args[2];
    let path = PathBuf::from(file_path)
        .canonicalize()
        .map_err(|_| "Couldn't get abs path for schema".to_string())?;
    let abs_file_path = path
        .to_str()
        .ok_or("Couldn't get abs path for schema".to_string())?;

    let fs = RealFileSystem {};

    match command {
        Command::Format => format(&fs, abs_file_path)?,
        Command::Codegen => {
            let code = codegen(&fs, abs_file_path)?;
            println!("{}", code);
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, collections::HashMap, io};

    use crate::{codegen, format, FileSystem};

    pub struct MockFileSystem {
        pub files: RefCell<HashMap<String, String>>,
    }

    impl FileSystem for MockFileSystem {
        fn read_to_string(&self, path: &str) -> io::Result<String> {
            self.files
                .borrow()
                .get(path)
                .cloned()
                .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "File not found"))
        }

        fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
            self.files
                .borrow_mut()
                .insert(path.to_string(), content.to_string());
            Ok(())
        }

        fn resolve(
            &self,
            _: &str,
            target_file: &str,
        ) -> Result<String, Box<dyn std::error::Error>> {
            Ok(target_file.to_string())
        }
    }

    #[test]
    fn test_parse_example_sdl() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([
                (
                    "another-schema.memorix".to_string(),
                    r#"
Config {
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}
Type { abc: u32 }
Cache {
  number_of_messages: {
    payload: u32
    public: [get]
  }
  messages: {
    key: string
    payload: string
  }
}
"#
                    .to_string(),
                ),
                (
                    "schema.memorix".to_string(),
                    r#"
Config {
  import: [
    "another-schema.memorix"
  ]
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}

Namespace MessageService {
    Enum {
      Operation {
        START
        STOP
      }
    }
    Cache {
        message: {
            key: string
            payload: {
            id: string
            sender_id: UserId
            recipient_id: UserId
            content: string
            timestamp: u64
            }
            public: [get]
        }
    }

    PubSub {
        new_message: {
            key: UserId
            payload: {
            message_id: string
            recipient_id: UserId
            }
            public: [subscribe]
        }
    }

    Task {
        message_processing_tasks: {
            payload: {
                message_id: string
                processing_type: string
                priority: u32
            }
            queue_type: "Fifo"
        }
    }
}

NamespaceDefaults {
  cache_ttl: "3600"
}

Cache {
  user_profile: {
    key: u32
    payload: {
      id: u32
      name: string
      email: string
    }
    public: [get]
  }
  user_session: {
    key: string
    payload: {
      user_id: u32
      session_token: string
      expiry: u64
    }
    ttl: env(USER_SESSION_TTL)
  }
}
  
Namespace UserService {
  NamespaceDefaults {
    cache_ttl: "7200"
    task_queue_type: env(USER_SERVICE_TASK_QUEUE_TYPE)
  }

  Type {
    UserId: u64
  }

  Cache {
    user_profile: {
      key: UserId
      payload: {
        id: UserId
        name: string
        email: string
      }
      public: [get]
    }
    user_session: {
      key: string
      payload: {
        user_id: UserId
        session_token: string
        expiry: u64
      }
      ttl: env(USER_SESSION_TTL)
    }
  }

  PubSub {
    user_activity: {
      key: UserId
      payload: {
        user_id: UserId
        action: string
        timestamp: u64
      }
      public: [subscribe]
    }
  }

  Task {
    user_registration_tasks: {
      payload: {
        user_id: UserId
        email: string
        registration_date: u64
      }
      public: [enqueue get_len]
    }
  }

}

"#
                    .to_string(),
                ),
            ])),
        };
        insta::assert_snapshot!(codegen(&mock_fs, "schema.memorix")?);

        Ok(())
    }

    #[test]
    fn test_fail_nicely() -> Result<(), Box<dyn std::error::Error>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
Config {
  export: {
    engine: Redis(env(REDIS_URL))
    files: [      {
        language: typescript
        path: "memorix.generated.ts"
      }]
  }
}

Type {
  a: [u64 u32]
}

"#
                .to_string(),
            )])),
        };

        let result = codegen(&mock_fs, "schema.memorix");
        let err = result.unwrap_err();
        insta::assert_snapshot!(err.to_string());

        Ok(())
    }
    #[test]
    fn test_format() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
    Config {
      export: {
        engine: Redis(env(REDIS_URL))
        files: [      {
            language: typescript
            path: "memorix.generated.ts"
          }]
      }
    }
    Namespace Science {
      Cache {
        how_many_atoms: {
          payload: u32
          key: u64
        }
      }
      PubSub {
        how_many_atoms: {
          payload: u32
        }
      }
    }
    Type {
      a: [u64      ]
    }
    Namespace Rocket {
      Cache {
        launched: {
          payload: boolean
        }
      }
      Enum {
        Operation {
          START
          STOP
        }
        Color {
          RED
          GREEN
          BLUE
        }
      }
      PubSub {
        launched: {
          payload: boolean
        }
      }
    }

    "#
                .to_string(),
            )])),
        };
        format(&mock_fs, "schema.memorix")?;
        let formatted_schema = mock_fs.read_to_string("schema.memorix")?;
        insta::assert_snapshot!(formatted_schema);

        Ok(())
    }
}
