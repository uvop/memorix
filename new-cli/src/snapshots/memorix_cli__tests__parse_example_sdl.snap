---
source: src/main.rs
assertion_line: 386
expression: "codegen(&mock_fs, \"schema.memorix\")?"
---
[
  {
    "global_namespace": {
      "type_object_items": [],
      "modified_namespace": {
        "defaults": {
          "cache_ttl": null,
          "task_queue_type": null
        },
        "type_items": [
          [
            "abc",
            "U32"
          ]
        ],
        "enum_items": [],
        "cache_items": [
          [
            "number_of_messages",
            {
              "key": null,
              "payload": "U32",
              "ttl": null,
              "expose": [
                "Get",
                "Set",
                "Delete"
              ]
            }
          ],
          [
            "messages",
            {
              "key": "String",
              "payload": "String",
              "ttl": null,
              "expose": [
                "Get",
                "Set",
                "Delete"
              ]
            }
          ]
        ],
        "pubsub_items": [],
        "task_items": []
      }
    },
    "namespaces": []
  },
  {
    "global_namespace": {
      "type_object_items": [
        [
          "InlineCachePayloadUserProfile",
          {
            "properties": [
              [
                "id",
                "U32"
              ],
              [
                "name",
                "String"
              ],
              [
                "email",
                "String"
              ]
            ]
          }
        ],
        [
          "InlineCachePayloadUserSession",
          {
            "properties": [
              [
                "user_id",
                "U32"
              ],
              [
                "session_token",
                "String"
              ],
              [
                "expiry",
                "U64"
              ]
            ]
          }
        ]
      ],
      "modified_namespace": {
        "defaults": {
          "cache_ttl": {
            "String": "3600"
          },
          "task_queue_type": null
        },
        "type_items": [
          [
            "abc",
            "U32"
          ]
        ],
        "enum_items": [],
        "cache_items": [
          [
            "user_profile",
            {
              "key": "U32",
              "payload": {
                "Reference": "InlineCachePayloadUserProfile"
              },
              "ttl": null,
              "expose": [
                "Get",
                "Set",
                "Delete"
              ]
            }
          ],
          [
            "user_session",
            {
              "key": "String",
              "payload": {
                "Reference": "InlineCachePayloadUserSession"
              },
              "ttl": {
                "Env": "USER_SESSION_TTL"
              },
              "expose": [
                "Get",
                "Set",
                "Delete"
              ]
            }
          ],
          [
            "number_of_messages",
            {
              "key": null,
              "payload": "U32",
              "ttl": null,
              "expose": [
                "Get"
              ]
            }
          ]
        ],
        "pubsub_items": [],
        "task_items": []
      }
    },
    "namespaces": [
      [
        "MessageService",
        {
          "type_object_items": [
            [
              "InlineCachePayloadMessage",
              {
                "properties": [
                  [
                    "id",
                    "String"
                  ],
                  [
                    "sender_id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "recipient_id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "content",
                    "String"
                  ],
                  [
                    "timestamp",
                    "U64"
                  ]
                ]
              }
            ],
            [
              "InlinePubSubPayloadNewMessage",
              {
                "properties": [
                  [
                    "message_id",
                    "String"
                  ],
                  [
                    "recipient_id",
                    {
                      "Reference": "UserId"
                    }
                  ]
                ]
              }
            ],
            [
              "InlineTaskPayloadMessageProcessingTasks",
              {
                "properties": [
                  [
                    "message_id",
                    "String"
                  ],
                  [
                    "processing_type",
                    "String"
                  ],
                  [
                    "priority",
                    "U32"
                  ]
                ]
              }
            ]
          ],
          "modified_namespace": {
            "defaults": {
              "cache_ttl": null,
              "task_queue_type": null
            },
            "type_items": [],
            "enum_items": [
              [
                "Operation",
                [
                  "START",
                  "STOP"
                ]
              ]
            ],
            "cache_items": [
              [
                "message",
                {
                  "key": "String",
                  "payload": {
                    "Reference": "InlineCachePayloadMessage"
                  },
                  "ttl": null,
                  "expose": [
                    "Get",
                    "Set",
                    "Delete"
                  ]
                }
              ]
            ],
            "pubsub_items": [
              [
                "new_message",
                {
                  "key": {
                    "Reference": "UserId"
                  },
                  "payload": {
                    "Reference": "InlinePubSubPayloadNewMessage"
                  },
                  "expose": [
                    "Publish",
                    "Subscribe"
                  ]
                }
              ]
            ],
            "task_items": [
              [
                "message_processing_tasks",
                {
                  "key": null,
                  "payload": {
                    "Reference": "InlineTaskPayloadMessageProcessingTasks"
                  },
                  "queue_type": {
                    "String": "Fifo"
                  },
                  "expose": [
                    "Enqueue",
                    "Dequeue",
                    "Empty",
                    "GetLen"
                  ]
                }
              ]
            ]
          }
        }
      ],
      [
        "UserService",
        {
          "type_object_items": [
            [
              "InlineCachePayloadUserProfile",
              {
                "properties": [
                  [
                    "id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "name",
                    "String"
                  ],
                  [
                    "email",
                    "String"
                  ]
                ]
              }
            ],
            [
              "InlineCachePayloadUserSession",
              {
                "properties": [
                  [
                    "user_id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "session_token",
                    "String"
                  ],
                  [
                    "expiry",
                    "U64"
                  ]
                ]
              }
            ],
            [
              "InlinePubSubPayloadUserActivity",
              {
                "properties": [
                  [
                    "user_id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "action",
                    {
                      "Optional": "String"
                    }
                  ],
                  [
                    "timestamp",
                    "U64"
                  ]
                ]
              }
            ],
            [
              "InlineTaskPayloadUserRegistrationTasks",
              {
                "properties": [
                  [
                    "user_id",
                    {
                      "Reference": "UserId"
                    }
                  ],
                  [
                    "email",
                    "String"
                  ],
                  [
                    "registration_date",
                    "U64"
                  ]
                ]
              }
            ]
          ],
          "modified_namespace": {
            "defaults": {
              "cache_ttl": {
                "String": "7200"
              },
              "task_queue_type": {
                "Env": "USER_SERVICE_TASK_QUEUE_TYPE"
              }
            },
            "type_items": [
              [
                "UserId",
                "U64"
              ]
            ],
            "enum_items": [],
            "cache_items": [
              [
                "user_profile",
                {
                  "key": {
                    "Reference": "UserId"
                  },
                  "payload": {
                    "Reference": "InlineCachePayloadUserProfile"
                  },
                  "ttl": null,
                  "expose": [
                    "Get",
                    "Set",
                    "Delete"
                  ]
                }
              ],
              [
                "user_session",
                {
                  "key": "String",
                  "payload": {
                    "Reference": "InlineCachePayloadUserSession"
                  },
                  "ttl": {
                    "Env": "USER_SESSION_TTL"
                  },
                  "expose": [
                    "Get",
                    "Set",
                    "Delete"
                  ]
                }
              ]
            ],
            "pubsub_items": [
              [
                "user_activity",
                {
                  "key": {
                    "Reference": "UserId"
                  },
                  "payload": {
                    "Reference": "InlinePubSubPayloadUserActivity"
                  },
                  "expose": [
                    "Publish",
                    "Subscribe"
                  ]
                }
              ]
            ],
            "task_items": [
              [
                "user_registration_tasks",
                {
                  "key": null,
                  "payload": {
                    "Reference": "InlineTaskPayloadUserRegistrationTasks"
                  },
                  "queue_type": null,
                  "expose": [
                    "Enqueue",
                    "Dequeue",
                    "Empty",
                    "GetLen"
                  ]
                }
              ]
            ]
          }
        }
      ]
    ]
  }
]
